language/evm/move-to-yul/src/yul_functions.rs

```diff
@@ -106,12 +120,259 @@ fn make_yule_def(name: &str, body: &str) -> String {
 
 // The Yul functions supporting the compilation scheme.
 functions! {
+// -------------------------------------------------------------------------------------------
+// Abort
 Abort: "(code) {
     revert(0, 0) // TODO: convention to store code
 }",
 AbortBuiltin: "() {
     $Abort(sub(0, 1))
 }" dep Abort,
+
+// -------------------------------------------------------------------------------------------
+// Memory
+
+// Allocates memory of size.
+// TODO: add some memory recovery (e.g. over free lists), and benchmark against the current
+//   arena style.
+Malloc: "(size) -> offs {
+    offs := mload(${MEM_SIZE_LOC})
+    mstore(${MEM_SIZE_LOC}, add(offs, size))
+}",
+
+// Frees memory of size
+Free: "(offs, size) {
+}",
+
+// Makes a pointer, using the lowest bit to indicate whether it is for storage or memory.
+MakePtr: "(is_storage, offs) -> ptr {
+  ptr := or(is_storage, shl(offs, 1))
+}",
+
+// Returns true if this is a storage  pointer.
+IsStoragePtr: "(ptr) -> b {
+  b := and(ptr, 0x1)
+}",
+
+// Returns the offset of this pointer.
+OffsetPtr: "(ptr) -> offs {
+  offs := shr(ptr, 1)
+}",
+
+// Takes an offset and splits it into a word and a bit offset.
+ToWordOffs: "(offs) -> word_offs, bit_offs {
+  word_offs := shr(offs, 5)
+  bit_offs := shl(and(offs, 0x1F), 3)
+}",
+
+// Make a unique key into storage, where word can have full 32 byte size, and type
+// indicates the kind of the key given as a byte. This uses keccak256 to fold
+// value and type into a unique storage key.
+StorageKey: "(type, word) -> key {
+  mstore(${SCRATCH1_LOC}, word)
+  mstore(${SCRATCH2_LOC}, type)
+  key := keccak256(${SCRATCH1_LOC}, 33)
+}",
+
+// Indexes pointer by offset.
+IndexPtr: "(ptr, offs) -> new_ptr {
+  new_ptr := $MakePtr($IsStoragePtr(ptr), add($OffsetPtr(ptr), offs))
+}" dep MakePtr dep IsStoragePtr dep OffsetPtr,
+
+// Loads u8 from pointer.
+LoadU8: "(ptr) -> val {
+  let offs := $OffsetPtr(ptr)
+  switch $IsStoragePtr(ptr)
+  case 0 {
+    val := $MemoryLoadU8(offs)
+  }
+  default {
+    val := $StorageLoadU8(offs)
+  }
+}" dep OffsetPtr dep IsStoragePtr dep MemoryLoadU8 dep StorageLoadU8,
+
+// Loads u8 from memory offset.
+MemoryLoadU8: "(offs) -> val {
+  val := and(mload(offs), ${MAX_U8})
+}",
+
+// Loads u8 from storage offset.
+StorageLoadU8: "(offs) -> val {
+  let word_offs, bit_offs := $ToWordOffs(offs)
+  let key := $StorageKey(${CONTINUOUS_STORAGE_TYPE}, word_offs)
+  val := and(shr(sload(key), bit_offs), ${MAX_U8})
```

Comment:
Are we using little endian within the word here, where smaller bit offset stores less significant bit? So bit offset 0 is actually the least significant bit of the 256 bits. This seems a bit counterintuitive. If we want to use big endian instead we should `shr` by `(256 - bit_offs - 8)`.

Same confusion for the other storage load and store functions.

Reply:
This was written under the wrong assumption of little endian, and is fixed in PR #79